### PR TITLE
pserver etcd client

### PR DESCRIPTION
**Pull request by jacquesqiao**
fix: https://github.com/PaddlePaddle/Paddle/issues/2515

### CODE REVIEW

File: go/pserver/cclient/cclient.go [hunk 1]
```diff
@@ -100,9 +105,39 @@ func paddle_new_pserver_client(addrs *C.char, selected int) C.paddle_pserver_cli
...
 
 //export paddle_new_etcd_pserver_client
-func paddle_new_etcd_pserver_client(etcd_addr *C.char) C.paddle_pserver_client {
-	// TODO(helin): fault tolerant pserver client using etcd.
-	panic("not implemented.")
+func paddle_new_etcd_pserver_client(etcd_addr *C.char, pserver_path *C.char, selected int) C.paddle_pserver_client {
+	if etcd_addr == "" {
+		return C.PSERVER_ERROR
+	}
+	etcd_addrs := strings.Split(etcd_addr, ",")
```
Comment (helinwang) [inline go/pserver/cclient/cclient.go]: Sorry I should have mentioned this earlier, the cgo code are supposed to be just a thin layer. Because it's just a interface from C to Go. If too much code here, then when we want to use Pserver from Go to Go, we need to duplicate code. Another reason is Go's test does not work well with cgo, it's supposed to be tested from C side. It's much easier to write test for Go.

You can create a new file in pserver package. E.g., etcd.go, and implement this interface: https://github.com/PaddlePaddle/Paddle/blob/develop/go/pserver/client.go#L26
And this interface:
https://github.com/PaddlePaddle/Paddle/blob/develop/go/pserver/client.go#L15

So cgo just need to do:
```Go
e := pserver.NewEtcd(addr) // we can call it NewEtcd, or some other name.
c := pserver.NewClient(e, len(as), e)
return add(c)
```

File: go/pserver/cclient/cclient.go [hunk 2]
```diff
@@ -100,9 +105,39 @@ func paddle_new_pserver_client(addrs *C.char, selected int) C.paddle_pserver_cli
...
+		if err != nil {
+			log.Errorf("get %s error %v", pserver_path, err)
+		}
+		kvs := resp.Kvs
+		if len(kvs) == 0 {
+			log.Infoln("Waiting for pservers register, sleeping")
+			time.Sleep(time.Second)
+			continue
+		}
+		ps_addr := string(kvs[0].Value)
```
Comment (typhoonzero) [inline go/pserver/cclient/cclient.go]: From design doc:  https://github.com/PaddlePaddle/Paddle/tree/develop/doc/design/cluster_train#trainer-process-1

`ps_addr` is not fetched from one etcd key. Addresses are read from `/ps/<id>` which stands for IP address for each pserver, so `ps_addr` need to concat all these addresses.

And each trainer must "watch" pserver nodes hand handle events of pserver creation and loss.

File: go/pserver/cclient/cclient.go (hunk 1 above)
Comment (jacquesqiao) [inline go/pserver/cclient/cclient.go]: Thank you for the comment! I will add a separate interface to do this.

File: go/pserver/cclient/cclient.go [hunk 3]
```diff
@@ -100,9 +105,39 @@ func paddle_new_pserver_client(addrs *C.char, selected int) C.paddle_pserver_cli
...
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   etcd_addrs,
+		DialTimeout: etcdTimeout,
+	})
+	if err != nil {
+		log.Errorln(err)
+		return err
+	}
+	retryTimes := defaultRetryTimes
+	for retryTimes < 0 {
```
Comment (jacquesqiao) [inline go/pserver/cclient/cclient.go]: Ok, I will optimize this strategy

File: go/pserver/cclient/cclient.go [hunk 4]
```diff
@@ -100,9 +105,39 @@ func paddle_new_pserver_client(addrs *C.char, selected int) C.paddle_pserver_cli
...
+		return C.PSERVER_ERROR
+	}
+	etcd_addrs := strings.Split(etcd_addr, ",")
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   etcd_addrs,
+		DialTimeout: etcdTimeout,
+	})
+	if err != nil {
+		log.Errorln(err)
+		return err
```
Comment (Yancey1989) [inline go/pserver/cclient/cclient.go]: The return type is `C.paddle_pserver_client`, I think the function would panic if client connect etcd failed.

File: go/pserver/etcd_lister.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+const DefaultEtcdTimeout time.Duration = time.Second * time.Duration(5)
+
+type pserverEtcdLister struct {
+	client    *clientv3.Client
+	timeout   time.Duration
+	endpoints []string
+}
+
+// read ps desired number from etcd.
+func(p pserverEtcdLister) desired() int {
```
Comment (typhoonzero) [inline go/pserver/etcd_lister.go]: `p pserverEtcdLister` ->` p *pserverEtcdLister` same as `func(p pserverEtcdLister) List() []Server`

Comment (typhoonzero) [inline go/pserver/etcd_lister.go]: Can put https://github.com/PaddlePaddle/Paddle/blob/develop/go/pserver/service.go#L106 as a function and call it directly.

File: go/pserver/service.go [hunk 6]
```diff
@@ -21,6 +21,10 @@ type ElementType int
 const (
 	AlreadyInitialized = "pserver already initialized"
 	Uninitialized      = "pserver not fully initialized"
+	// PsDesired is etcd path for store desired pserver count
+	DefaultPsDesiredPath = "/ps_desired"
```
Comment (typhoonzero) [inline go/pserver/service.go]: Just a note: etcdv3 uses only key-values, thought the key looks like a path.

File: go/pserver/etcd_lister.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+			time.Sleep(p.timeout)
+			continue
+		}
+
+		log.Debugf("Get psDesired number: %d\n", psDesired)
+		return psDesired
+	}
+}
+
+func(p pserverEtcdLister) List() []Server {
```
Comment (typhoonzero) [inline go/pserver/etcd_lister.go]: In https://github.com/PaddlePaddle/Paddle/pull/2551 , trainers need to discover pservers and keep watch pservers using etcd's watch. I think you need to discuss with @Yancey1989 to see where wo implement the "etcd watcher"

Comment (jacquesqiao) [inline go/pserver/etcd_lister.go]: Thanks for reminding! I will discuss with your today~~

File: go/pserver/etcd_lister.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	DefaultEtcdTimeout time.Duration = time.Second * time.Duration(5)
```
Comment (helinwang) [inline go/pserver/etcd_lister.go]: `5 * time.Second` is fine. Constants in Go is very well designed: https://blog.golang.org/constants

File: go/pserver/service.go (hunk 6 above)
Comment (jacquesqiao) [inline go/pserver/service.go]: ok~

File: go/pserver/etcd_lister.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+	}
+}
+
+func(p *pserverEtcdLister) List() []Server {
+	psDesired := p.desired()
+
+	servers := make([]Server, psDesired)
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+		for i := 0; i < psDesired; i++ {
```
Comment (helinwang) [inline go/pserver/etcd_lister.go]: This can be batched into a single transaction, here is an related example: https://github.com/PaddlePaddle/Paddle/blob/develop/go/master/etcd_client.go#L121
Something like:
```Go
resp, err := e.client.Txn(ctx).If().Then(get0, get1, get2, get3).Commit()
```
Feel free to improve it later, not a merge blocker.

File: go/pserver/etcd_lister.go [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+				log.Infof("Waiting for ps addr registered ...")
+				time.Sleep(p.timeout)
+				continue
+			}
+
+			psAddr := string(resp.Kvs[0].Value)
+			// TODO(Longfei) check the ps address
+			if  psAddr == "" {
+				cancel()
+				log.Infof("Get psKey = %s,  psAddr is null illegal", psKey, psAddr)
```
Comment (helinwang) [inline go/pserver/etcd_lister.go]: Only one "%s", but two variable. `go vet` is a good tool for detecting this kind of error.

File: go/pserver/etcd_lister.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+			if err != nil {
+				cancel()
+				log.Infof("Get psKey= %s error, %v", psKey, err)
+				time.Sleep(p.timeout)
+				continue
+			}
+			kvs := resp.Kvs
+			if len(kvs) == 0 {
+				log.Infof("Waiting for ps addr registered ...")
+				time.Sleep(p.timeout)
```
Comment (helinwang) [inline go/pserver/etcd_lister.go]: `cancel()` need to be called.

File: go/pserver/etcd_lister.go [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+	"github.com/coreos/etcd/clientv3"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	DefaultEtcdTimeout time.Duration = time.Second * time.Duration(5)
+)
+
+type pserverEtcdLister struct {
```
Comment (helinwang) [inline go/pserver/etcd_lister.go]: The name could be more concise if it's something like `etcdLister`, the package is already `pserver`, people read it in their minds as: pserver etcd lister. If it's named as `pserverEtcdLister`, it's long, and does not read good: pserver pserver etcd lister.

File: go/pserver/etcd_lister.go [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+			servers[i].Index = i
+			servers[i].Addr = psAddr
+		}
+		cancel()
+		break
+	}
+	return servers
+}
+
+func NewEtcdAddrLister(endpoints string) (Lister, int) {
```
Comment (helinwang) [inline go/pserver/etcd_lister.go]: It's less abstraction (interface is an abstraction) if the return type is `PserverEtcdLister` (I think it's a good idea to make `pserverEtcdLister` public, perhaps rename to `EtcdLister`, and `desired` public), rather than `Lister`.

The benefit of less abstraction is not very obvious in the beginning, maybe the article below could explain more. I will give a not-so-good-example: At least it's easier to jump to the definition when not returning an interface (editor jumps to the function definition rather than interface definition).

A good general rule in Go would be: [accept interfaces, return structs](https://medium.com/@cep21/preemptive-interface-anti-pattern-in-go-54c18ac0668a)

File: go/pserver/client_test.go [hunk 14]
```diff
@@ -125,3 +174,20 @@ func TestClientFull(t *testing.T) {
...
+	initNativeClient()
+	servers := make([]pserver.Server, numPserver)
+	for i := 0; i < numPserver; i++ {
+		servers[i] = pserver.Server{Index: i, Addr: ":" + strconv.Itoa(port[i])}
+	}
+	c1 := pserver.NewClient(lister(servers), len(servers), selector(true))
+	ClientTest(t, c1)
+}
+
+func TestEtcdClient(t *testing.T) {
```
Comment (jacquesqiao) [inline go/pserver/client_test.go]: ok

File: go/pserver/etcd_lister.go (hunk 8 above)
Comment (jacquesqiao) [inline go/pserver/etcd_lister.go]: done

File: go/pserver/etcd_lister.go (hunk 11 above)
Comment (jacquesqiao) [inline go/pserver/etcd_lister.go]: done

File: go/pserver/etcd_lister.go (hunk 10 above)
Comment (jacquesqiao) [inline go/pserver/etcd_lister.go]: ok, done, thanx~

File: go/pserver/etcd_lister.go (hunk 12 above)
Comment (jacquesqiao) [inline go/pserver/etcd_lister.go]: done, thanks for this very valuable suggestion!

File: go/pserver/cclient/cclient.go [hunk 15]
```diff
@@ -100,9 +100,11 @@ func paddle_new_pserver_client(addrs *C.char, selected int) C.paddle_pserver_cli
 }
 
 //export paddle_new_etcd_pserver_client
-func paddle_new_etcd_pserver_client(etcd_addr *C.char) C.paddle_pserver_client {
-	// TODO(helin): fault tolerant pserver client using etcd.
-	panic("not implemented.")
+func paddle_new_etcd_pserver_client(etcd_addr *C.char, selected int) C.paddle_pserver_client {
```
Comment (jacquesqiao) [inline go/pserver/cclient/cclient.go]: I add a TODO to optimize this part and will learn how to use etcd lock to sync status.

File: go/pserver/cclient/cclient.go (hunk 4 above)
Comment (jacquesqiao) [inline go/pserver/cclient/cclient.go]: done

File: go/pserver/cclient/cclient.go (hunk 2 above)
Comment (jacquesqiao) [inline go/pserver/cclient/cclient.go]: Done the first part, but I have not implemented the watcher because now the pserver client cannot change to new pserver when running, I have added a TODO and will implment this function in the following pr. Thank you!

File: go/pserver/cclient/cclient.go [hunk 16]
```diff
@@ -100,9 +100,11 @@ func paddle_new_pserver_client(addrs *C.char, selected int) C.paddle_pserver_cli
...
+	addr := C.GoString(etcd_addr)
```
Comment (jacquesqiao) [inline go/pserver/cclient/cclient.go]: done

File: go/master/etcd_client.go [hunk 17]
```diff
@@ -98,7 +98,7 @@ func (e *EtcdClient) Save(state []byte) error {
 			// We lost the master lock and can not acquire
 			// it back, it means some other master is
 			// already started. We don't want cluster
-			// managment system to kill the master server
+			// management system to kill the master server
```
Comment (jacquesqiao) [inline go/master/etcd_client.go]: :)

File: go/pserver/client_test.go [hunk 18]
```diff
@@ -124,3 +154,21 @@ func TestClientFull(t *testing.T) {
...
+	servers := make([]pserver.Server, numPserver)
+	for i := 0; i < numPserver; i++ {
+		servers[i] = pserver.Server{Index: i, Addr: ":" + strconv.Itoa(pserverClientPorts[i])}
+	}
+	c1 := pserver.NewClient(lister(servers), len(servers), selector(true))
+	ClientTest(t, c1)
+}
+
+//TODO(Qiao: tmperary disable etcdClient test for dependency of etcd)
+func EtcdClient(t *testing.T) {
```
Comment (helinwang) [inline go/pserver/client_test.go]: Please make sure to disable it before merging into develop branch. We need to make sure test always runs correctly, since other developers need to run test as well.

File: go/pserver/etcd_cclient.go [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	DefaultEtcdTimeout time.Duration = 5 * time.Second
```
Comment (helinwang) [inline go/pserver/etcd_cclient.go]: Just FYI `DefaultEtcdTimeout = 5 * time.Second` could work as well.

File: go/pserver/etcd_cclient.go [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+)
+
+type EtcdCClient interface {
```
Comment (helinwang) [inline go/pserver/etcd_cclient.go]: I know it's hard to choose a name, since `EtcdClient` is already used ;)
However, `EtcdCClient` may not be a good name, because C Client means it's for C. But here is not really related to C (Go can use it as well). Maybe `ClientEtcd`?

File: go/pserver/etcd_cclient.go [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+	Desired() int
+	List() []Server
+}
+
+// TODO(Longfei)
+// 1. add watcher to watch the change state of pservers)
+// 1. add etcd lock)
+type EtcdCClientImpl struct {
```
Comment (helinwang) [inline go/pserver/etcd_cclient.go]: It's a very common pattern in Java: create interface, and write the implementation, this is due to the explicit interface in Java. It's not required for Go: Go has implicit interface. Please see: https://medium.com/@cep21/preemptive-interface-anti-pattern-in-go-54c18ac0668a Unnecessary interface only adds complication to the program.

I think we need to stick to "accept interface, return struct" unless there is good reason not to.
In this case we need to remove `type EtcdCClient interface`, rename `EtcdCClientImpl` to `EtcdCClient`, and make `func NewEtcdCClient(endpoints string) (*EtcdCClient, error)` return pointer to the struct.

File: go/pserver/etcd_cclient.go [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+func (p *EtcdCClientImpl) List() []Server {
+	psDesired := p.Desired()
+
+	servers := make([]Server, psDesired)
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+		for i := 0; i < psDesired; i++ {
+			psKey := PsPath + strconv.Itoa(i)
+			log.Debugf("checking %s", psKey)
+			resp, err := p.client.Get(ctx, psKey)
```
Comment (helinwang) [inline go/pserver/etcd_cclient.go]: 是不是改成
```go
ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
resp, err := p.client.Get(ctx, psKey)
cancel()
if err != nil {
...
```
更清楚些？

Comment (jacquesqiao) [inline go/pserver/etcd_cclient.go]: 好的~

Comment (jacquesqiao) [inline go/pserver/etcd_cclient.go]: 这个地方还不能这么改，因为是在一个循环里边，只能最后cancel。否则下次循环到这里就失败了

File: go/pserver/etcd_client.go [hunk 23]
```diff
@@ -64,11 +71,11 @@ func (e *EtcdClient) Register() (int, error) {
 		log.Debugf("inited client to %s", e.etcdEndpoints)
 		break
 	}
-	// init /ps_desired using transaction, for multiple pservers may want to write
+	// init /ps/desired using transaction, for multiple pservers may want to write
```
Comment (typhoonzero) [inline go/pserver/etcd_client.go]: We use `ps_desired` in design doc: https://github.com/PaddlePaddle/Paddle/tree/develop/doc/design/cluster_train#parameter-server-process-1

File: go/pserver/client/etcd_client.go [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/PaddlePaddle/Paddle/go/pserver"
+	"github.com/coreos/etcd/clientv3"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	DefaultEtcdTimeout time.Duration = 5 * time.Second
+)
+
+// TODO
```
Comment (typhoonzero) [inline go/pserver/client/etcd_client.go]: Todo comment styles must be like `// TODO: ...`

File: go/pserver/client/etcd_client.go [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+// 1. add watcher to watch the change state of pservers)
```
Comment (typhoonzero) [inline go/pserver/client/etcd_client.go]: Comment for exported structs and functions must be like `// EtcdClient ...`

Comment (typhoonzero) [inline go/pserver/client/etcd_client.go]: You can use [gometalinter](https://github.com/alecthomas/gometalinter) to check the code styles.

File: go/pserver/client/etcd_client.go [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+		}
+
+		psDesired, err := strconv.Atoi(string(resp.Kvs[0].Value))
+		if err != nil {
+			log.Errorf("psDesired %s invalid %v", psDesired, err)
+			time.Sleep(p.timeout)
+			continue
+		}
+
+		log.Debugf("Get psDesired number: %d\n", psDesired)
```
Comment (typhoonzero) [inline go/pserver/client/etcd_client.go]: Logrus formated log seems don't need a `\n`

File: go/pserver/client/etcd_client.go [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+		return psDesired
+	}
+}
+
+func (p *EtcdClient) List() []Server {
```
Comment (typhoonzero) [inline go/pserver/client/etcd_client.go]: Exported function must have comments.

File: go/pserver/client/etcd_client.go [hunk 28]
```diff
@@ -0,0 +1,124 @@
...
+
+		psDesired, err := strconv.Atoi(string(resp.Kvs[0].Value))
+		if err != nil {
+			log.Errorf("psDesired %s invalid %v", psDesired, err)
+			time.Sleep(p.timeout)
+			continue
+		}
+
+		log.Debugf("Get psDesired number: %d\n", psDesired)
+		return psDesired
```
Comment (typhoonzero) [inline go/pserver/client/etcd_client.go]: Break and return.

File: go/pserver/client/etcd_client.go [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+
+	servers := make([]Server, psDesired)
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+		for i := 0; i < psDesired; i++ {
+			psKey := pserver.PsPath + strconv.Itoa(i)
+			log.Debugf("checking %s", psKey)
+			resp, err := p.client.Get(ctx, psKey)
+			if err != nil {
+				cancel()
```
Comment (typhoonzero) [inline go/pserver/client/etcd_client.go]: Not sure how the `cancel()` should be used. Should always place behind `Get`?

File: go/pserver/client/etcd_client.go [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+func (p *EtcdClient) List() []Server {
+	psDesired := p.Desired()
+
+	servers := make([]Server, psDesired)
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+		for i := 0; i < psDesired; i++ {
+			psKey := pserver.PsPath + strconv.Itoa(i)
+			log.Debugf("checking %s", psKey)
+			resp, err := p.client.Get(ctx, psKey)
```
Comment (typhoonzero) [inline go/pserver/client/etcd_client.go]: Pserver clients or trainers may need to "watch" pserver node changes(like pserver down  and restarted and recoverd on another node). But this may be in another PR, add some TODO here.

File: go/pserver/client/etcd_client.go [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+// TODO
+// 1. add watcher to watch the change state of pservers)
+// 1. add etcd lock)
+type EtcdClient struct {
+	client    *clientv3.Client
+	timeout   time.Duration
+	endpoints []string
+}
+
+// read ps desired number from etcd.
```
Comment (Yancey1989) [inline go/pserver/client/etcd_client.go]: `// Desired  ...`

File: go/pserver/client/etcd_client.go [hunk 32]
```diff
@@ -0,0 +1,124 @@
...
+
+func NewEtcd(endpoints string) (*EtcdClient, error) {
+	ep := strings.Split(endpoints, ",")
+	timeout := DefaultEtcdTimeout
+	var cli *clientv3.Client
+	var err error
+	for {
+		cli, err = clientv3.New(clientv3.Config{
+			Endpoints:   ep,
+			DialTimeout: timeout,
```
Comment (Yancey1989) [inline go/pserver/client/etcd_client.go]: If we always use the `DefaultEtcdTimeout`, maybe we do not need the local variable `timeout`

File: go/pserver/client/etcd_client.go [hunk 33]
```diff
@@ -0,0 +1,124 @@
...
+		break
+	}
+	return servers
+}
+
+func NewEtcd(endpoints string) (*EtcdClient, error) {
+	ep := strings.Split(endpoints, ",")
+	timeout := DefaultEtcdTimeout
+	var cli *clientv3.Client
+	var err error
```
Comment (Yancey1989) [inline go/pserver/client/etcd_client.go]: Maybe we don't need the local variable `err`, and we can initialize this with `cli, err := clientv3.New(...)`

File: go/pserver/client/etcd_client.go [hunk 34]
```diff
@@ -0,0 +1,124 @@
...
+			servers[i].Index = i
+			servers[i].Addr = psAddr
+		}
+		cancel()
+		break
+	}
+	return servers
+}
+
+func NewEtcd(endpoints string) (*EtcdClient, error) {
```
Comment (Yancey1989) [inline go/pserver/client/etcd_client.go]: With the implementation of `NewEtcd`, it will never return the error with `not nil`, so I think we can return the error if we initialize the etc client failed.

File: go/pserver/etcd_cclient.go (hunk 22 above)
Comment (helinwang) [inline go/pserver/etcd_cclient.go]: @jacquesqiao 下一个循环貌似又创建了一个新的ctx?
```go
for ... {
  ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
  resp, err := p.client.Get(ctx, psKey)
  cancel()
}
```

File: go/pserver/etcd_client.go (hunk 23 above)
Comment (jacquesqiao) [inline go/pserver/etcd_client.go]: done

File: go/pserver/client/etcd_client.go (hunk 24 above)
Comment (jacquesqiao) [inline go/pserver/client/etcd_client.go]: done

File: go/pserver/client/etcd_client.go (hunk 25 above)
Comment (jacquesqiao) [inline go/pserver/client/etcd_client.go]: done

File: go/pserver/client/etcd_client.go (hunk 26 above)
Comment (jacquesqiao) [inline go/pserver/client/etcd_client.go]: removed

File: go/pserver/client/etcd_client.go (hunk 28 above)
Comment (jacquesqiao) [inline go/pserver/client/etcd_client.go]: done

File: go/pserver/client/etcd_client.go (hunk 27 above)
Comment (jacquesqiao) [inline go/pserver/client/etcd_client.go]: done

File: go/pserver/client/etcd_client.go (hunk 30 above)
Comment (jacquesqiao) [inline go/pserver/client/etcd_client.go]: ok, I will do this in next pr.

File: go/pserver/client/etcd_client.go (hunk 29 above)
Comment (jacquesqiao) [inline go/pserver/client/etcd_client.go]: done

File: go/pserver/client/etcd_client.go (hunk 31 above)
Comment (jacquesqiao) [inline go/pserver/client/etcd_client.go]: done

File: go/pserver/client/etcd_client.go (hunk 33 above)
Comment (jacquesqiao) [inline go/pserver/client/etcd_client.go]: the `cli` will be used outside the for loop, so I have to declare it here, so I can't use `cli, err := clientv3.New(...)`..

File: go/pserver/client/etcd_client.go (hunk 32 above)
Comment (jacquesqiao) [inline go/pserver/client/etcd_client.go]: done

File: go/pserver/etcd_client.go [hunk 35]
```diff
@@ -136,7 +143,7 @@ func (e *EtcdClient) registerPserverEtcd(ctx context.Context) (int, error) {
 	_, err := concurrency.NewSTM(e.etcdClient, func(c concurrency.STM) error {
 		registered := false
 		for i := 0; i < e.desired; i++ {
-			psKey := "/ps/" + strconv.Itoa(i)
+			psKey := PsPath + strconv.Itoa(i)
 			log.Debugf("checking %s", psKey)
 			ps := c.Get(psKey)
 			log.Debugf("got value (%s) for key: %s", ps, psKey)
```
Comment (jacquesqiao) [inline go/pserver/etcd_client.go]: This is a common style with other code in paddle, so I will not change it this time.

File: go/pserver/etcd_cclient.go (hunk 22 above)
Comment (jacquesqiao) [inline go/pserver/etcd_cclient.go]: done

File: go/pserver/etcd_lister.go (hunk 9 above)
Comment (jacquesqiao) [inline go/pserver/etcd_lister.go]: ok

File: go/pserver/client/etcd_client.go (hunk 34 above)
Comment (jacquesqiao) [inline go/pserver/client/etcd_client.go]: remove unused err msg

File: go/pserver/etcd_client.go (hunk 23 above)
Comment (typhoonzero) [inline go/pserver/etcd_client.go]: Comment is not updated

File: go/pserver/etcd_client.go [hunk 36]
```diff
@@ -77,8 +84,8 @@ func (e *EtcdClient) Register() (int, error) {
 		}
 		break
 	}
-	// TODO: when implementing extending or reducing pservers, /ps_desired is
-	// changed, then we need to watch /ps_desired node for events. For now, just
+	// TODO: when implementing extending or reducing pservers, /ps/desired is
+	// changed, then we need to watch /ps/desired node for events. For now, just
```
Comment (typhoonzero) [inline go/pserver/etcd_client.go]: Comment is not upated

Comment (jacquesqiao) [inline go/pserver/etcd_client.go]: fix done
